File: watch/src/main/java/com/example/wearbabymonitor/MainActivity.java
package com.example.wearbabymonitor;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.gms.wearable.Wearable;

import java.nio.charset.StandardCharsets;
import java.util.UUID;

public final class MainActivity extends Activity {
    private static final int REQUEST_PERMISSIONS = 10;
    private static final int TEST_ACK_POLLS = 20;
    private static final long TEST_ACK_POLL_MS = 250L;

    private TextView status;
    private Button sensitivity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        status = new TextView(this);
        status.setText("Stopped");
        status.setTextSize(16f);
        status.setGravity(Gravity.CENTER);
        silenceView(status);

        Button start = button("START", this::ensurePermissionsAndStart);
        Button stop = button("STOP", this::stopMonitor);
        Button test = button("TEST PHONE", this::sendEndToEndTest);
        Button recalibrate = button("RECALIBRATE ROOM", this::recalibrateRoom);
        sensitivity = button(sensitivityLabel(), this::cycleSensitivity);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setGravity(Gravity.CENTER_HORIZONTAL);
        layout.setPadding(26, 26, 26, 48);
        silenceView(layout);
        layout.addView(status);
        layout.addView(start);
        layout.addView(stop);
        layout.addView(test);
        layout.addView(recalibrate);
        layout.addView(sensitivity);

        ScrollView scrollView = new ScrollView(this);
        scrollView.setFillViewport(true);
        scrollView.setVerticalScrollBarEnabled(true);
        scrollView.setSmoothScrollingEnabled(true);
        silenceView(scrollView);
        scrollView.addView(
                layout,
                new ScrollView.LayoutParams(
                        ScrollView.LayoutParams.MATCH_PARENT,
                        ScrollView.LayoutParams.WRAP_CONTENT
                )
        );
        setContentView(scrollView);
        refreshStatus();
    }

    @Override
    protected void onResume() {
        super.onResume();
        refreshStatus();
    }

    private Button button(String text, Runnable action) {
        Button button = new Button(this);
        button.setText(text);
        button.setMinHeight(56);
        button.setAllCaps(false);
        silenceView(button);
        button.setOnClickListener(view -> action.run());
        return button;
    }

    private void silenceView(View view) {
        view.setSoundEffectsEnabled(false);
        view.setHapticFeedbackEnabled(false);
    }

    private String sensitivityLabel() {
        String value = getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE)
                .getString(NoiseMonitorService.KEY_SENSITIVITY, "medium");
        return "SENSITIVITY: " + value.toUpperCase(java.util.Locale.US);
    }

    private void cycleSensitivity() {
        String current = getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE)
                .getString(NoiseMonitorService.KEY_SENSITIVITY, "medium");
        String next = "medium";
        if ("medium".equals(current)) next = "high";
        else if ("high".equals(current)) next = "low";
        getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE)
                .edit().putString(NoiseMonitorService.KEY_SENSITIVITY, next).apply();
        sensitivity.setText(sensitivityLabel());
        status.setText("Sensitivity applies next time monitoring starts");
    }

    private void refreshStatus() {
        android.content.SharedPreferences prefs = getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE);
        boolean active = prefs.getBoolean(NoiseMonitorService.KEY_MONITORING_ACTIVE, false);
        String state = prefs.getString(NoiseMonitorService.KEY_MONITORING_STATE, active ? "monitoring" : "stopped");
        if (!active) {
            status.setText("Stopped");
        } else if ("calibrating".equals(state)) {
            status.setText("Calibrating — keep the room quiet");
        } else {
            status.setText("Monitoring");
        }
    }

    private void recalibrateRoom() {
        boolean active = getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE)
                .getBoolean(NoiseMonitorService.KEY_MONITORING_ACTIVE, false);
        if (!active) {
            status.setText("Start monitoring first");
            return;
        }
        Intent intent = new Intent(this, NoiseMonitorService.class);
        intent.setAction(NoiseMonitorService.ACTION_RECALIBRATE);
        startForegroundService(intent);
        status.setText("Recalibrating — keep the room quiet");
    }

    private void ensurePermissionsAndStart() {
        java.util.ArrayList<String> missing = new java.util.ArrayList<>();
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            missing.add(Manifest.permission.RECORD_AUDIO);
        }
        if (android.os.Build.VERSION.SDK_INT >= 33
                && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
            missing.add(Manifest.permission.POST_NOTIFICATIONS);
        }

        if (missing.isEmpty()) {
            startMonitor();
        } else {
            requestPermissions(missing.toArray(new String[0]), REQUEST_PERMISSIONS);
        }
    }

    private void startMonitor() {
        boolean alreadyActive = getSharedPreferences(NoiseMonitorService.PREFS, MODE_PRIVATE)
                .getBoolean(NoiseMonitorService.KEY_MONITORING_ACTIVE, false);
        if (alreadyActive) {
            status.setText("Already monitoring");
            return;
        }
        startForegroundService(new Intent(this, NoiseMonitorService.class));
        status.setText("Calibrating — keep the room quiet");
    }

    private void stopMonitor() {
        sendStatus(false, "user_stopped");
        stopService(new Intent(this, NoiseMonitorService.class));
        status.setText("Stopped");
    }

    private void sendEndToEndTest() {
        String alertId = UUID.randomUUID().toString();
        getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
                .edit()
                .remove(WatchMessageListenerService.KEY_LAST_ACK)
                .apply();
        status.setText("Testing…");

        Wearable.getNodeClient(this).getConnectedNodes().addOnSuccessListener(nodes -> {
            if (nodes.isEmpty()) {
                status.setText("Phone disconnected");
                return;
            }
            byte[] payload = Protocol.encodeAlert(alertId, Protocol.ALERT_TYPE_TEST, 0.0);
            for (com.google.android.gms.wearable.Node node : nodes) {
                Wearable.getMessageClient(this).sendMessage(node.getId(), Protocol.NOISE_PATH, payload);
            }
            waitForAck(alertId);
        }).addOnFailureListener(error -> status.setText("Test failed"));
    }

    private void waitForAck(String alertId) {
        Thread thread = new Thread(() -> {
            boolean acknowledged = false;
            for (int poll = 0; poll < TEST_ACK_POLLS && !Thread.currentThread().isInterrupted(); poll++) {
                if (alertId.equals(lastAck())) {
                    acknowledged = true;
                    break;
                }
                try {
                    Thread.sleep(TEST_ACK_POLL_MS);
                } catch (InterruptedException interrupted) {
                    Thread.currentThread().interrupt();
                }
            }
            final boolean result = acknowledged;
            runOnUiThread(() -> status.setText(result ? "Phone confirmed" : "Test failed"));
        }, "baby-monitor-test-ack");
        thread.start();
    }

    private String lastAck() {
        return getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
                .getString(WatchMessageListenerService.KEY_LAST_ACK, null);
    }

    private void sendStatus(boolean monitoring, String state) {
        String text = "battery=-1,connected=true,monitoring=" + monitoring
                + ",state=" + state + ",time=" + System.currentTimeMillis();
        byte[] payload = text.getBytes(StandardCharsets.UTF_8);
        Wearable.getNodeClient(this).getConnectedNodes().addOnSuccessListener(nodes -> {
            for (com.google.android.gms.wearable.Node node : nodes) {
                Wearable.getMessageClient(this).sendMessage(node.getId(), Protocol.STATUS_PATH, payload);
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode != REQUEST_PERMISSIONS) return;
        boolean granted = grantResults.length > 0;
        for (int result : grantResults) {
            granted &= result == PackageManager.PERMISSION_GRANTED;
        }
        if (granted) startMonitor();
        else status.setText("Permissions required");
    }
}

File: watch/src/main/java/com/example/wearbabymonitor/MainActivity.kt
package com.example.wearbabymonitor

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.Gravity
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.tasks.Tasks
import com.google.android.gms.wearable.Wearable
import java.util.UUID
import java.util.concurrent.TimeUnit

class MainActivity : Activity() {
    private lateinit var status: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        status = TextView(this).apply {
            text = "Stopped"
            textSize = 16f
            gravity = Gravity.CENTER
        }

        val start = Button(this).apply {
            text = "START"
            setOnClickListener { ensurePermissionsAndStart() }
        }
        val stop = Button(this).apply {
            text = "STOP"
            setOnClickListener {
                sendStoppedStatus()
                stopService(Intent(this@MainActivity, NoiseMonitorService::class.java))
                status.text = "Stopped"
            }
        }
        val test = Button(this).apply {
            text = "TEST PHONE"
            setOnClickListener { sendEndToEndTest() }
        }

        setContentView(LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
            setPadding(18, 18, 18, 18)
            addView(status)
            addView(start)
            addView(stop)
            addView(test)
        })
    }

    private fun ensurePermissionsAndStart() {
        val missing = mutableListOf<String>()
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            missing += Manifest.permission.RECORD_AUDIO
        }
        if (Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            missing += Manifest.permission.POST_NOTIFICATIONS
        }

        if (missing.isNotEmpty()) {
            ActivityCompat.requestPermissions(this, missing.toTypedArray(), REQUEST_PERMISSIONS)
        } else {
            startMonitor()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_PERMISSIONS &&
            grantResults.isNotEmpty() && grantResults.all { it == PackageManager.PERMISSION_GRANTED }
        ) {
            startMonitor()
        } else {
            status.text = "Permissions required"
        }
    }

    private fun startMonitor() {
        ContextCompat.startForegroundService(
            this,
            Intent(this, NoiseMonitorService::class.java)
        )
        status.text = "Calibrating"
    }

    private fun sendEndToEndTest() {
        val alertId = UUID.randomUUID().toString()
        getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
            .edit()
            .remove(WatchMessageListenerService.KEY_LAST_ACK)
            .apply()
        status.text = "Testing…"

        Thread {
            var acknowledged = false
            try {
                for (attempt in 0 until TEST_SEND_ATTEMPTS) {
                    val nodes = Tasks.await(
                        Wearable.getNodeClient(this).connectedNodes,
                        TEST_TASK_TIMEOUT_SECONDS,
                        TimeUnit.SECONDS
                    )
                    for (node in nodes) {
                        Tasks.await(
                            Wearable.getMessageClient(this).sendMessage(
                                node.id,
                                Protocol.NOISE_PATH,
                                Protocol.encodeAlert(
                                    id = alertId,
                                    type = Protocol.ALERT_TYPE_TEST,
                                    rms = 0.0
                                )
                            ),
                            TEST_TASK_TIMEOUT_SECONDS,
                            TimeUnit.SECONDS
                        )
                    }

                    for (poll in 0 until TEST_ACK_POLLS) {
                        Thread.sleep(TEST_ACK_POLL_MS)
                        if (lastAck() == alertId) {
                            acknowledged = true
                            break
                        }
                    }
                    if (acknowledged) break
                }
            } catch (_: Exception) {
                acknowledged = false
            }

            runOnUiThread {
                status.text = if (acknowledged) "Phone confirmed" else "Test failed"
            }
        }.apply {
            name = "baby-monitor-e2e-test"
            start()
        }
    }

    private fun sendStoppedStatus() {
        val payload = "battery=-1,connected=true,monitoring=false,state=user_stopped,time=${System.currentTimeMillis()}"
            .encodeToByteArray()
        Wearable.getNodeClient(this).connectedNodes.addOnSuccessListener { nodes ->
            nodes.forEach { node ->
                Wearable.getMessageClient(this)
                    .sendMessage(node.id, Protocol.STATUS_PATH, payload)
            }
        }
    }

    private fun lastAck(): String? =
        getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
            .getString(WatchMessageListenerService.KEY_LAST_ACK, null)

    companion object {
        private const val REQUEST_PERMISSIONS = 10
        private const val TEST_SEND_ATTEMPTS = 4
        private const val TEST_ACK_POLLS = 5
        private const val TEST_ACK_POLL_MS = 300L
        private const val TEST_TASK_TIMEOUT_SECONDS = 3L
    }
}
